### DISCUSSION TITLE
Difference between torch one_hot and ignite to_onehot

### DISCUSSION THREAD

**Discussion by jeffydc** [pytorch/ignite#1550] [Q&A] [answered]
As far as I know, pytorch return torch.int64 while ignite return torch.unit8 and pytorch add `num_classes` in the last dimension, ignite adds `num_classes` in the `dim=1`.
Besides those differences, is there any other differences or any other specific use case?
```py
import torch
from ignite.utils import to_onehot, manual_seed

manual_seed(666)

x = torch.randint(10, (2, 5))

print(to_onehot(x, 10), to_onehot(x, 10).shape)
print(torch.nn.functional.one_hot(x, 10), torch.nn.functional.one_hot(x, 10).shape)


x = torch.randint(10, (5,))

print(to_onehot(x, 10), to_onehot(x, 10).shape)
print(torch.nn.functional.one_hot(x, 10), torch.nn.functional.one_hot(x, 10).shape)
```
Output:
```py
tensor([[[1, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 1, 0, 0, 1],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 1, 0],
         [0, 0, 1, 0, 0]],

        [[0, 0, 1, 0, 0],
         [0, 0, 0, 0, 0],
         [1, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 1, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 1, 0, 0, 1],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0]]], dtype=torch.uint8) torch.Size([2, 10, 5])
tensor([[[1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
         [0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
         [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
         [0, 0, 0, 0, 1, 0, 0, 0, 0, 0]],

        [[0, 0, 1, 0, 0, 0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
         [1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
         [0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0, 0, 0, 1, 0, 0]]]) torch.Size([2, 5, 10])
tensor([[0, 0, 0, 0, 0, 1, 0, 0, 0, 0],
        [0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
        [0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 0, 1, 0, 0, 0, 0]], dtype=torch.uint8) torch.Size([5, 10])
tensor([[0, 0, 0, 0, 0, 1, 0, 0, 0, 0],
        [0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
        [0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 0, 1, 0, 0, 0, 0]]) torch.Size([5, 10])
```

**Comment by vfdev-5** [comment, marked as answer]
> Besides those differences, is there any other differences or any other specific use case?

I think that's it. uint8 vs long is about memory impact. We put `num_classes` to `dim=1` such that targets with shapes like `(B, H, W)` can be easily transformed to one-hot targets `(B, C, H, W)`.

**Reply by jeffydc** [reply]
I see! Thank you @vfdev-5 I can use one or another depending on the shape i want.
